File: src/homebox_companion/core/app_preferences.py
"""Application preferences management.

This module provides storage and retrieval of application-level preferences
that can be customized by users through the Settings UI.

Settings are persisted to {data_dir}/app_preferences.json
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from loguru import logger
from pydantic import BaseModel, Field

from homebox_companion.core.config import settings


def _get_prefs_path() -> Path:
    """Get the app preferences file path based on settings.data_dir."""
    return Path(settings.data_dir) / "app_preferences.json"


class AppPreferences(BaseModel):
    """Application preferences that can be customized via UI.

    These preferences allow users to override environment-variable defaults
    without needing to modify their deployment configuration.
    """

    # Connection settings
    homebox_url_override: str | None = Field(
        default=None,
        description="Override for HOMEBOX_URL environment variable",
    )
    image_quality_override: str | None = Field(
        default=None,
        description="Override for HBC_IMAGE_QUALITY (low, medium, high)",
    )

    # Behavior settings
    duplicate_detection_enabled: bool = Field(
        default=True,
        description="Enable duplicate detection by serial number",
    )
    show_token_usage: bool = Field(
        default=False,
        description="Display AI token usage counts in detection results",
    )

    # Enrichment settings
    enrichment_enabled: bool = Field(
        default=False,
        description="Enable AI-powered product specification enrichment",
    )
    enrichment_auto_enrich: bool = Field(
        default=False,
        description="Automatically enrich items after AI detection",
    )
    enrichment_cache_ttl_hours: int = Field(
        default=24,
        ge=1,
        le=168,
        description="Cache duration for enrichment results (hours)",
    )

    # Web search settings for enrichment
    search_provider: str = Field(
        default="none",
        description="Web search provider for enrichment (none, tavily, google_cse, searxng)",
    )
    search_tavily_api_key: str | None = Field(
        default=None,
        description="Tavily API key (get at https://tavily.com)",
    )
    search_google_api_key: str | None = Field(
        default=None,
        description="Google Cloud API key for Custom Search",
    )
    search_google_engine_id: str | None = Field(
        default=None,
        description="Google Programmable Search Engine ID (cx)",
    )
    search_searxng_url: str | None = Field(
        default=None,
        description="SearXNG instance URL (e.g., https://searx.example.com)",
    )

    # Custom retailer domains for price fetching
    enrichment_retailer_domains: list[str] = Field(
        default_factory=list,
        description="Additional retailer domains to fetch prices from (e.g., microcenter.com)",
    )


def load_app_preferences() -> AppPreferences:
    """Load application preferences from file.

    Returns:
        AppPreferences instance with values from file, or defaults.
    """
    prefs_file = _get_prefs_path()
    if not prefs_file.exists():
        return AppPreferences()

    try:
        file_data = json.loads(prefs_file.read_text(encoding="utf-8"))
        return AppPreferences.model_validate(file_data)
    except (json.JSONDecodeError, Exception) as e:
        logger.warning(f"Invalid app preferences file, using defaults: {e}")
        return AppPreferences()


def save_app_preferences(prefs: AppPreferences) -> None:
    """Save application preferences to file.

    Args:
        prefs: The preferences to save.
    """
    prefs_file = _get_prefs_path()
    prefs_file.parent.mkdir(parents=True, exist_ok=True)
    data = prefs.model_dump()
    prefs_file.write_text(json.dumps(data, indent=2), encoding="utf-8")
    logger.info(f"App preferences saved to {prefs_file}")


def get_effective_homebox_url(prefs: AppPreferences) -> str:
    """Get the effective Homebox URL, considering override.

    Args:
        prefs: The app preferences.

    Returns:
        The Homebox URL to use (override or from settings).
    """
    from homebox_companion.core.config import settings

    if prefs.homebox_url_override:
        return prefs.homebox_url_override
    return settings.api_url


def get_effective_image_quality(prefs: AppPreferences) -> str:
    """Get the effective image quality, considering override.

    Args:
        prefs: The app preferences.

    Returns:
        The image quality to use (override or from settings).
    """
    from homebox_companion.core.config import settings

    if prefs.image_quality_override:
        return prefs.image_quality_override
    return settings.image_quality

File: frontend/src/lib/api/appPreferences.ts
/**
 * App Preferences API endpoints
 */

import { request } from './client';

// =============================================================================
// TYPES
// =============================================================================

export interface AppPreferencesResponse {
	homebox_url_override: string | null;
	image_quality_override: string | null;
	duplicate_detection_enabled: boolean;
	show_token_usage: boolean;
	enrichment_enabled: boolean;
	enrichment_auto_enrich: boolean;
	enrichment_cache_ttl_hours: number;
	// Web search settings
	search_provider: string;
	search_tavily_api_key: string | null;
	search_google_api_key: string | null;
	search_google_engine_id: string | null;
	search_searxng_url: string | null;
	// Custom retailer domains
	enrichment_retailer_domains: string[];
	// Effective values
	effective_homebox_url: string;
	effective_image_quality: string;
	image_quality_options: string[];
}

export interface AppPreferencesInput {
	homebox_url_override: string | null;
	image_quality_override: string | null;
	duplicate_detection_enabled: boolean;
	show_token_usage: boolean;
	enrichment_enabled: boolean;
	enrichment_auto_enrich: boolean;
	enrichment_cache_ttl_hours: number;
	// Web search settings
	search_provider: string;
	search_tavily_api_key: string | null;
	search_google_api_key: string | null;
	search_google_engine_id: string | null;
	search_searxng_url: string | null;
	// Custom retailer domains
	enrichment_retailer_domains: string[];
}

// =============================================================================
// API FUNCTIONS
// =============================================================================

/**
 * Get current app preferences
 */
export const getAppPreferences = () =>
	request<AppPreferencesResponse>('/settings/app-preferences');

/**
 * Update app preferences
 */
export const updateAppPreferences = (prefs: AppPreferencesInput) =>
	request<AppPreferencesResponse>('/settings/app-preferences', {
		method: 'PUT',
		body: JSON.stringify(prefs),
	});
